File: sdk/go-guest/examples/blob.go
package export_tests_local_blobs_api

import (
	wit "go.bytecodealliance.org/pkg/wit/types"
	"runtime"
	raw "wit_component/latent_blob_blob"
	blob "wit_component/lsf/blob"
)

var invocations uint32

func Run(which uint32, _ string, handle uint64) uint64 {
	invocations++
	if invocations != 1 {
		panic("blob fixture must start with fresh guest state")
	}
	// Raw generated access is used only by the stale-host-handle negative cases.
	if which == 2 || which == 4 {
		if which == 2 {
			handle = raw.Create("text/plain", wit.Some[uint64](0)).Ok()
			if !raw.Close(handle).Ok() {
				panic("writer did not close")
			}
		}
		r := raw.Write(handle, 0, []byte{})
		if r.IsErr() && r.Err().Tag() == raw.BlobErrorInvalidState {
			return 10
		}
		if r.IsErr() && r.Err().Tag() == raw.BlobErrorPermissionDenied {
			return 11
		}
		panic("closed or foreign handle accepted")
	}
	if which == 5 {
		return raw.Create("text/plain", wit.Some[uint64](0)).Ok()
	}
	if which == 1 {
		// The real SDK cancellation case holds Create pending in the provider.
		// Rendezvous with four bounded application goroutines before that wait;
		// cancellation must reclaim their stacks without running Go defers.
		started := make(chan struct{}, 4)
		parked := make(chan struct{})
		for worker := 0; worker < 4; worker++ {
			go func() {
				started <- struct{}{}
				<-parked
			}()
		}
		for worker := 0; worker < 4; worker++ {
			<-started
		}
		runtime.Gosched()
		defer close(parked)
	}
	writer := blob.Create("text/plain", wit.Some[uint64](4)).Ok()
	if which == 1 {
		return 1
	} // charged until activation cleanup; no hidden task
	if writer.Write(0, []byte("data")).Ok() != 4 {
		panic("short blob write")
	}
	reader := blob.Open(writer.Seal().Ok()).Ok()
	chunk := reader.Read(0, 4).Ok()
	defer chunk.Close()
	if !reader.Close().Ok() {
		panic("reader did not close")
	}
	if which == 3 {
		return 3
	}
	bytes := chunk.Bytes().Ok()
	if string(bytes) != "data" {
		panic("blob bytes changed")
	}
	return uint64(len(bytes))
}

File: sdk/dotnet-guest/examples/blob.cs
using Lsf.Guest;
using Raw = ServiceWorld.wit.Imports.latent.blob.IBlobImports;
namespace ServiceWorld.wit.Exports.tests.localBlobs;
public class ApiExportsImpl : IApiExports
{
    private static ulong Failure(Result<ulong, Raw.BlobError> result) => !result.IsOk ? result.AsErr.Tag switch {
        Raw.BlobError.Tags.InvalidState => 10UL,
        Raw.BlobError.Tags.PermissionDenied => 11UL,
        _ => throw new System.InvalidOperationException("unexpected blob result") }
        : throw new System.InvalidOperationException("stale blob handle succeeded");
    public static ulong Run(uint which, string text, ulong handle)
    {
        if (which == 4) return Failure(Raw.Write(handle, 0, []));
        using var scope = new Scope();
        var writer = scope.Own(Blob.Create("text/plain", which is 2 or 5 ? 0UL : 4UL).AsOk);
        if (which == 1) return 1;
        if (which == 5) return writer.Consume(value => value);
        if (which == 2) {
            var stale = writer.Borrow(value => value);
            _ = Blob.Close(writer).AsOk;
            return Failure(Raw.Write(stale, 0, []));
        }
        _ = Blob.Write(writer, 0, [100, 97, 116, 97]).AsOk;
        var reference = Blob.Seal(writer).AsOk;
        var reader = scope.Own(Blob.Open(reference).AsOk);
        var chunk = scope.Own(Blob.Read(reader, 0, 4).AsOk);
        _ = Blob.Close(reader).AsOk;
        if (which == 3) return 3;
        var bytes = Blob.ChunkBytes(chunk).AsOk;
        if (which == 6) {
            var again = Blob.ChunkBytes(chunk);
            if (!again.IsOk && again.AsErr.Tag == Raw.BlobError.Tags.InvalidState) return 10;
            throw new System.InvalidOperationException("materialized a chunk twice");
        }
        return (ulong)bytes.Length;
    }
}
